packages/flutter/lib/src/widgets/splash.dart

```diff
@@ -0,0 +1,621 @@
+
+  /// If asserts are enabled, this value tracks whether the feature has been disposed.
+  ///
+  /// Ensures that [dispose] is only called once, and [paint] is not called afterward.
+  bool debugDisposed = false;
+
+  /// A (typically translucent) color used for this [Splash].
+  Color get color => _color;
+  Color _color;
+  set color(Color value) {
+    if (value == _color) {
+      return;
+    }
+    _color = value;
+    controller.markNeedsPaint();
+  }
+
+  /// A [ShapeBorder] that may optionally be applied to the [Splash].
+  ShapeBorder? get customBorder => _customBorder;
+  ShapeBorder? _customBorder;
+  set customBorder(ShapeBorder? value) {
+    if (value == _customBorder) {
+      return;
+    }
+    _customBorder = value;
+    controller.markNeedsPaint();
+  }
+
+  /// Draws a [Splash] on the provided [Canvas].
+  ///
+  /// The [transform] argument is the [Matrix4] transform that typically
+  /// shifts the coordinate space of the canvas to the space in which
+  /// the circle is to be painted.
+  ///
+  /// If a [customBorder] is provided, then it (along with the [textDirection])
+  /// will be used to create a clipping path.
+  ///
+  /// Otherwise, the [clipCallback] clips the splash to a [RRect] (created by
+  /// applying the [borderRadius] to its result).
+  ///
+  /// If both [customBorder] and [clipCallback] are null, no clipping takes place.
+  ///
+  /// For examples on how the function is used, see [InkSplash] and [InkRipple].
+  @protected
+  void paintCircle({
+    required Canvas canvas,
+    required Matrix4 transform,
+    required Paint paint,
+    required Offset center,
+    required double radius,
+    TextDirection? textDirection,
+    ShapeBorder? customBorder,
+    BorderRadius borderRadius = BorderRadius.zero,
+    RectCallback? clipCallback,
+  }) {
+    final Offset? originOffset = MatrixUtils.getAsTranslation(transform);
+    canvas.save();
+    if (originOffset == null) {
+      canvas.transform(transform.storage);
+    } else {
+      canvas.translate(originOffset.dx, originOffset.dy);
+    }
+    if (clipCallback != null) {
+      final Rect rect = clipCallback();
+      if (customBorder != null) {
+        canvas.clipPath(customBorder.getOuterPath(rect, textDirection: textDirection));
+      } else if (borderRadius != BorderRadius.zero) {
+        canvas.clipRRect(RRect.fromRectAndCorners(
+          rect,
+          topLeft: borderRadius.topLeft, topRight: borderRadius.topRight,
+          bottomLeft: borderRadius.bottomLeft, bottomRight: borderRadius.bottomRight,
+        ));
+      } else {
+        canvas.clipRect(rect);
+      }
+    }
+    canvas.drawCircle(center, radius, paint);
+    canvas.restore();
+  }
+
+  /// Draws an ink circle on the provided [Canvas].
+  @protected
+  @Deprecated(
+    'Use paintCircle instead. '
+    '"Splash effects" no longer rely on a MaterialInkController. '
+    'This feature was deprecated after v3.23.0-0.1.pre.',
+  )
+  void paintInkCircle({
```

Comment:
Is there any reason to keep this method at all, which is a utility method and born deprecated?

Reply:
It's only here because we have

```dart
typedef InteractiveInkFeature = Splash;
```

so removing the method could cause breakages.